### PR TITLE
[GDN] Conv1d Tree Traversal Optimization for EAGLE Speculative Decoding

### PR DESCRIPTION
**Pull request by yyq0210**
## Summary

  - Pre-compute `retrive_parent_token` in `build_tree_kernel_efficient` CUDA kernel, eliminating redundant O(seqlen) reconstruction in the conv1d Triton kernel
  - Add token-parallel conv1d kernel variant that processes all tree tokens concurrently instead of serially
  - ~12-17% speedup on conv1d verify path for GDN (Mamba-hybrid) models with EAGLE topk > 1

## Motivation

In EAGLE speculative decoding with tree verification (`topk > 1`), the `causal_conv1d_update` Triton kernel previously:
  1. Received `retrieve_next_token` + `retrieve_next_sibling` encodings
  2. **Reconstructed** the parent map in-kernel using O(seqlen) `tl.where` + `tl.sum` per lookup
  3. Wrote out `retrieve_parent_token` as OUTPUT for downstream SSM kernels

  This had two problems:
  - Wasted ALU: the parent map is trivially known at tree-build time
  - Serial token loop: inter-token dependencies (parent_map construction) prevented parallelism

### Phase A: Pre-compute parent_map in `build_tree_kernel_efficient`

  | File | Change |
  |------|--------|
  | `sgl-kernel/csrc/speculative/eagle_utils.cu` | `build_tree_efficient` and `build_tree_efficient_partial_packed` kernels now write `retrive_parent_token` directly during tree construction |
  | `sgl-kernel/include/sgl_kernel_ops.h` | Added `retrive_parent_token` parameter to declaration |
  | `sgl-kernel/csrc/common_extension.cc` | Updated op schema |
  | `sgl-kernel/csrc/common_extension_rocm.cc` | Updated op schema |
  | `sgl-kernel/csrc/common_extension_musa.cc` | Updated op schema |
  | `sgl-kernel/python/sgl_kernel/speculative.py` | Added parameter to Pythonwrapper |
  | `python/sglang/srt/speculative/eagle_utils.py` | Allocate and pass `retrive_parent_token` to kernel, return from `build_tree_kernel_efficient` |
  | `python/sglang/srt/speculative/eagle_info.py` | Added `retrieve_parent_token` field to `EagleVerifyInput` |
  | `python/sglang/srt/speculative/eagle_worker.py` | Pass new field |
  | `python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py` | Read pre-computed `retrieve_parent_token` from `spec_info` instead of allocating empty |

  ### Phase B: Token-parallel conv1d kernel

  | File | Change |
  |------|--------|
  | `python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py` | Added `_causal_conv1d_verify_token_parallel_kernel` with grid `(batch, cdiv(dim, BLOCK_N), seqlen)`; serial kernel simplified to read `retrieve_parent_token` as INPUT; dispatch logic selects token-parallel when conditions met |

  ### Dispatch Logic

  ```python
  # Token-parallel kernel used when:
  # 1. retrieve_parent_token is provided (tree mode, topk > 1)
  # 2. num_accept_tokens is None (verify phase, not accept/reject rollback)
  # 3. seqlen > 1 (multiple draft tokens)
  use_token_parallel = (
      retrieve_parent_token is not None
      and num_accept_tokens is None
      and seqlen > 1
  )
 ```

### Test Environment

  - **GPU**: NVIDIA H20*2
  - **PyTorch**: 2.x with Triton
  - **Test method**: JIT compilation for CUDA
  (`torch.utils.cpp_extension.load`), direct kernel invocation for Triton
  - **Benchmark method**: 100 warmup + 500 timed iterations, median reported

### CUDA Kernel Parent Map (TestParentMapCorrectness) — 15 cases         
                                                                               
  | Test | Parameters | Result |                                               
  |------|-----------|--------|                                                
  | parent_map_consistent_with_tree | topk={2,4} × depth={2,3,4} × bs={1,2} | ✓ PASSED (12 cases) |
  | parent_map_expected_values | topk={2,4} | ✓ PASSED (2 cases) |
  | nullptr_no_crash | empty tensor | ✓ PASSED (1 case) |

  **Verification method**: JIT-compiled `eagle_utils.cu` via
  `torch.utils.cpp_extension.load()`

  **Assertions**:
  - `parent[0] == -1` (root has no parent)
  - `next_token[p] == t` → `parent[t] == p`
  - `next_sibling[s] == t` → `parent[t] == parent[s]`
  - For complete k-ary tree: `parent[i] == (i-1) // topk`

  ---

  ### Token-Parallel Kernel (TestTokenParallelCorrectness) — 152 cases

  #### Serial vs Token-Parallel (bit-exact match) — 144 cases

  | Parameter | Values |
  |-----------|--------|
  | batch | 1, 2, 4 |
  | dim | 256, 512, 2048 |
  | seqlen | 5, 7, 9, 13 |
  | width | 3, 4 |
  | dtype | bfloat16, float32 |

  **Tolerance**: float32: `rtol=1e-5, atol=1e-5` | bfloat16: `rtol=1e-2, atol=5e-2`

  All 144 combinations: **✓ PASSED** (output + intermediate_conv_window match)

  #### Additional Tests — 8 cases

  | Test | Description | Result |
  |------|-------------|--------|
  | test_token_parallel_no_bias (×2) | width={3,4}, no bias term | ✓ PASSED |
  | test_token_parallel_linear_chain | topk=1 (parent[i]=i-1) | ✓ PASSED |
  | test_against_reference (×4) | vs pure-Python reference, width={3,4}×seqlen={3,5} | ✓ PASSED |
  | test_dispatch_uses_token_parallel | verify `causal_conv1d_update` dispatch logic | ✓ PASSED |

  ---

### Conv1d Kernel Microbenchmark

  | Config | Serial | Token-Parallel | Speedup |
  |--------|--------|----------------|---------|
  | batch=4, dim=2048, seqlen=9, width=4 | 0.062ms | 0.055ms | **1.12x** |
  | batch=8, dim=2048, seqlen=9, width=4 | 0.062ms | 0.055ms | **1.13x** |
  | batch=4, dim=4096, seqlen=13, width=4 | 0.064ms | 0.055ms | **1.17x** |
  | batch=8, dim=4096, seqlen=17, width=4 | 0.070ms | 0.063ms | **1.11x** |
  | batch=16, dim=2048, seqlen=9, width=4 | 0.062ms | 0.056ms | **1.12x** |

  **Typical config**: EAGLE topk=2, depth=3 → seqlen=9, conv1d width=4

  **Summary**: **11%–17% speedup** across all configurations tested.

  ### E2E Benchmark Results
  
  **Conv1d Kernel: Serial vs Token-Parallel (CUDA event timing)**
  
  | Config | Serial | Token-Parallel | Speedup |
  |--------|--------|----------------|---------|
  | batch=4, dim=2048, seqlen=9, width=4 | 0.0518ms | 0.0438ms | 1.18x |
  | batch=8, dim=2048, seqlen=9, width=4 | 0.0511ms | 0.0437ms | 1.17x |
  | batch=4, dim=4096, seqlen=13, width=4 | 0.0515ms | 0.0436ms | 1.18x |
  | batch=8, dim=4096, seqlen=17, width=4 | 0.0556ms | 0.0481ms | 1.16x |
  | batch=16, dim=2048, seqlen=9, width=4 | 0.0509ms | 0.0439ms | 1.16x |
  | batch=8, dim=2048, seqlen=17, width=4 | 0.0534ms | 0.0449ms | 1.19x |
  | batch=16, dim=4096, seqlen=13, width=4 | 0.0561ms | 0.0505ms | 1.11x |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #28565742377](https://github.com/sgl-project/sglang/actions/runs/28565742377)
Latest PR Test (Extra): :x: [Run #28565742283](https://github.com/sgl-project/sglang/actions/runs/28565742283)

### CODE REVIEW

**Comment by yyq0210** [comment]
@BBuf @yuan-luo  sorry to bother you. Could you help to review this pr？
